### ISSUE TITLE
Cannot list nodes at cluster scope when using HelmCluster

### ISSUE THREAD

**Issue by adammrozik** [dask/dask-kubernetes#290]
Hi!

I have successfully deployed Dask on my Namespace within the cluster using Helm. However, I have issues running a simple code like the one below:

```python
from dask_kubernetes import HelmCluster
cluster = HelmCluster(release_name="my-helm-release", namespace='my-namespace')
cluster.scale(5)  # specify number of workers explicitly
```
as it results in permission error:

```python
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"nodes is forbidden: User \"me@email.com\" cannot list resource \"nodes\" in API group \"\" at the cluster scope","reason":"Forbidden","details":{"kind":"nodes"},"code":403}

```

As this is currently not possible with my permissions, is there something else I can do? It seems weird that scaling a Helm deployment that is just within my Namespace would need to list resources at the cluster scope.

Any help would be greatly appreciated!

**Environment**:

- Dask version: 2020.12.0; dask-kubernetes 0.11.0
- Python version: 3.7.1
- Operating System: MAC
- Install method (conda, pip, source): pip

**Comment by jacobtomlinson** [comment]
> It seems weird that scaling a Helm deployment that is just within my Namespace would need to list resources at the cluster scope.

Absolutely! Thanks for raising this.

From skimming through the code it seems this might be coming from the `NodePort` section. But if you could provide the full traceback that would be helpful.

https://github.com/dask/dask-kubernetes/blob/058103f9ea4138a799e73236b875b82489550d4a/dask_kubernetes/helm.py#L134

My guess here is that you deployed the Helm chart with the `NodePort` option which exposes the Dask cluster on a random port on all nodes in the cluster. In order for us to connect to the cluster we need an IP address of one of the nodes to connect to, hence listing nodes to get addresses.

I wonder if there is another way we can get this information.

**Comment by adammrozik** [comment]
Thanks for the quick answer!

Yeah, this is exactly what I did. My only option to expose dask scheduler for the outside world is to create a NodePort. And you have managed to find exactly the line that is causing the issue. Full traceback:
```
Traceback (most recent call last):
  File "dask_kubernetes_example.py", line 3, in <module>
    cluster = HelmCluster(release_name="helm-release", namespace='my-namespace')
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/dask_kubernetes/helm.py", line 95, in __init__
    self.sync(self._start)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/distributed/deploy/cluster.py", line 183, in sync
    return sync(self.loop, func, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/distributed/utils.py", line 340, in sync
    raise exc.with_traceback(tb)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/distributed/utils.py", line 324, in f
    result[0] = yield future
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/dask_kubernetes/helm.py", line 110, in _start
    self.scheduler_comm = rpc(await self._get_scheduler_address())
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/dask_kubernetes/helm.py", line 124, in _get_scheduler_address
    nodes = await self.core_api.list_node()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/kubernetes_asyncio/client/api_client.py", line 189, in __call_api
    raise e
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/kubernetes_asyncio/client/api_client.py", line 186, in __call_api
    _request_timeout=_request_timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/kubernetes_asyncio/client/rest.py", line 197, in GET
    query_params=query_params))
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/kubernetes_asyncio/client/rest.py", line 187, in request
    raise ApiException(http_resp=r)
kubernetes_asyncio.client.exceptions.ApiException: (403)
Reason: Forbidden
HTTP response headers: <CIMultiDictProxy(...)>
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"nodes is forbidden: User \"me@email.com\" cannot list resource \"nodes\" in API group \"\" at the cluster scope","reason":"Forbidden","details":{"kind":"nodes"},"code":403}


ERROR:asyncio:Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f8deb7b2be0>
ERROR:asyncio:Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f8deb7b2208>
ERROR:asyncio:Unclosed connector
connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x7f8deb783ce0>, 3.273242308)]']
connector: <aiohttp.connector.TCPConnector object at 0x7f8deb7b2c50>
```

I have a static DNS that I can connect to and DNS addresses as well. Is there a way I could provide it?

Also, similar scaling can be done with `kubectl` without Node listing e.g. `kubectl scale --replicas 5 deployment/helm-release-worker`. Not sure what the CLI does underneath, but maybe we could do something similar?

**Comment by jacobtomlinson** [comment]
> I have a static DNS that I can connect to and Node IP addresses as well. Is there a way I could provide it?

Not currently. But we could definitely add an extra kwarg to pass this along. Perhaps something like `nodeport_host`. Then if that value is set we could skip the node listing.

Would you be interested in contributing this?

> Also, similar scaling can be done with kubectl without Node listing e.g. kubectl scale --replicas 5 deployment/helm-release-worker. Not sure what the CLI does underneath, but maybe we could do something similar?

Sure. The `HelmCluster` class does much more than scaling though.

One common use case is to simplify connecting a client to the cluster. In your case it is less useful as you do not have enough permissions, but one example workflow would be.

```python
from dask_kubernetes import HelmCluster
from dask.distributed import Client

cluster = HelmCluster(release_name="my-helm-release", namespace='my-namespace')
client = Client(cluster)
```

Here we have been able to connect a client to the cluster without having to know anything about how the networking is configured. We just pass the helm release name and it works out the rest.

But in order to do this for `NodePort` deployments we need to list the nodes. If you chose `ClusterIP` or `LoadBalancer` this wouldn't be necessary. But of course those options have their own constraints.

**Comment by adammrozik** [comment]
Ah, I understand now. Thanks for the detailed explanation!

Would gladly contribute in some spare time! :)
I have created a small PR -> please advise me if anything else needs to be done.

https://github.com/dask/dask-kubernetes/pull/291

**Comment by adammrozik** [comment]
@jacobtomlinson : Thanks for the help on this one! Do you know when could I expect a new package release with the above change?

**Comment by jacobtomlinson** [comment]
I'm planning to get something out this week

**Comment by jacobtomlinson** [comment]
Closing as fixed in #291
